internal/ecrconsumer/image_values_test.go

```diff
@@ -0,0 +1,325 @@
+		expectedMap map[string]interface{}
+	}{
+		{
+			Image{
+				Registry:   "foo",
+				Repository: "bar",
+				Tag:        "newTag",
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+				},
+				"image": map[string]interface{}{
+					"repository": "foo/bar",
+					"tag":        "baz",
+				},
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+				},
+				"image": map[string]interface{}{
+					"repository": "foo/bar",
+					"tag":        "newTag",
+				},
+			},
+		}, {
+			Image{
+				Registry:   "foo",
+				Repository: "bar",
+				Tag:        "newTag",
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+					"image": map[string]interface{}{
+						"repository": "foo/bar",
+						"tag":        "baz",
+					},
+				},
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+					"image": map[string]interface{}{
+						"repository": "foo/bar",
+						"tag":        "newTag",
+					},
+				},
+			},
+		}, {
+			Image{
+				Registry:   "foo",
+				Repository: "bar",
+				Tag:        "newTag",
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+				},
+			},
+			map[string]interface{}{
+				"apples": "delicious",
+				"oranges": map[string]interface{}{
+					"taste": "delicious",
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		update(test.inputMap, test.newImage)
+		assert.Equal(t, test.expectedMap, test.inputMap)
+	}
+}
+
+func TestParseImage(t *testing.T) {
+	var tests = []struct {
+		repo     string
+		tag      string
+		expected *Image
+	}{
+		{
+			"foo/bar",
+			"baz",
+			&Image{
+				Registry:   "foo",
+				Repository: "bar",
+				Tag:        "baz",
+			},
+		}, {
+			"foo-bar",
+			"baz",
+			nil,
+		},
+	}
+	for _, test := range tests {
+		img, _ := parseImage(test.repo, test.tag)
+		assert.Equal(t, test.expected, img)
+	}
+}
+
+func TestWithImage(t *testing.T) {
+	expectedImg := Image{
+		Registry:   "bar",
+		Repository: "foo",
+		Tag:        "new-tag",
+	}
+
+	rls := v1alpha1.HelmRelease{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "HelmRelease",
+			APIVersion: "apiVersion: helmreleases.k8s.wattpad.com/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "example-microservice",
+		},
+		Spec: v1alpha1.HelmReleaseSpec{
+			ReleaseName: "example-release",
+			Chart: v1alpha1.ChartSpec{
+				Repository: "wattpad.s3.amazonaws.com/helm-charts",
+				Path:       "microservice",
+				Revision:   "HEAD",
+			},
+			Values: map[string]interface{}{
+				"image": map[string]interface{}{
+					"repository": "bar/foo",
+					"tag":        "baz",
+				},
+			},
+		},
+	}
+
+	outputRls := WithImage(expectedImg, rls)
```

Comment:
Couple thoughts on this test case. It's an important one, since the goal of this PR was to implement `WithImage`.

The overall behaviour we want to test is that `WithImage` finds and sets the desired image tag if there's a matching repo, and does nothing if there's no match. So we should have a test for both scenarios.

Second, we should avoid re-implementing the business logic we're trying to test, in the test case. `WithImage` calls `update` -> `getImagePath` & `table`, which are called again in the test case.

We know what our test data looks like, so we should take advantage of that:

```
outputRls := WithImage(expectedImg, rls)
outputImg := outputRls.Spec.Values["image"].(map[string]interface{})
assert.Equal(t, expectedImg.Tag, outputImg["tag"].(string)
```

For the "no image repo matches desired image" test case, we can assert that the original and modified HelmRelease are the same. Something like this should work:

```
// same setup as original TestWithImage
outputRls := WithImage(expectedImg, rls)
assert.Exactly(t, rls, outputRls)
```